src/main/java/graphql/nadel/engine/ArtificialFieldUtils.java

```diff
@@ -14,32 +14,47 @@
 import static graphql.language.Field.newField;
 
 /**
- * Interfaces and unions require that __typename be put on queries so we can work out what type they are on he other side
+ * Interfaces and unions require that __typename be put on queries so we can work out what type they are on he other side.
+ * We also add __typename field when all fields were deleted from a selection set to avoid producing non valid graphql
+ * with empty selection sets.
  */
 @Internal
 public class ArtificialFieldUtils {
 
     private static final String UNDERSCORE_TYPENAME = Introspection.TypeNameMetaFieldDef.getName();
+    private static final String TYPE_NAME_ALIAS_PREFIX_FOR_EMPTY_SELECTION_SETS = "empty_selection_set_";
+    public static final String TYPE_NAME_ALIAS_PREFIX_FOR_INTERFACES_AND_UNIONS = "type_hint_";
 
     public static Field maybeAddUnderscoreTypeName(NadelContext nadelContext, Field field, GraphQLOutputType fieldType) {
```

Comment:
similar question to the other: do we actually need to modify this method or can we add a second one which handles only the empty selection set part?

Reply:
replying to this in the other thread